### PR TITLE
COMPAT: Emit warning when groupby by a tuple

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/18314

### CODE REVIEW

**Comment by TomAugspurger** [comment]
cc @toobaz

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -2850,7 +2850,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
     elif isinstance(key, BaseGrouper):
         return key, [], obj
 
-    # Everything which is not a list is a key (including tuples):
+    tuple_as_list = isinstance(key, tuple) and key not in obj
```
Comment (toobaz) [inline pandas/core/groupby.py]: The only disadvantage I see with this approach is that
```python
pd.DataFrame(1, index=range(3), columns=pd.MultiIndex.from_product([[1, 2], [3,4]])).groupby((7, 8)).mean()
```
will raise ``KeyError: 7`` while  ``KeyError: (7,8)`` would be more correct. Do you think

```python
isinstance(key, tuple) and key not in obj and set(key).issubset(obj)
```
is too expensive?

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -2727,6 +2727,23 @@ def test_empty_dataframe_groupby(self):
...
+                           'b': [1, 2, 2, 2], 'c': [1, 1, 1, 1]})
+        with tm.assert_produces_warning(FutureWarning) as w:
+            df[['a', 'b', 'c']].groupby(('a', 'b')).c.mean()
+
+        assert "Interpreting tuple 'by' as a list" in str(w[0].message)
+
+        with tm.assert_produces_warning(FutureWarning) as w:
+            df[['a', 'b', 'c']].groupby(('a', 'b')).c.mean()
+
+        assert "Interpreting tuple 'by' as a list" in str(w[0].message)
```
Comment (toobaz) [inline pandas/tests/groupby/test_groupby.py]: Isn't this the same as above?

File: pandas/core/groupby.py (hunk 1 above)
Comment (jreback) [reply pandas/core/groupby.py]: this change would be ok

File: doc/source/whatsnew/v0.22.0.txt [hunk 3]
```diff
@@ -202,6 +202,9 @@ Deprecations
 - ``Series.from_array`` and ``SparseSeries.from_array`` are deprecated. Use the normal constructor ``Series(..)`` and ``SparseSeries(..)`` instead (:issue:`18213`).
 - ``DataFrame.as_matrix`` is deprecated. Use ``DataFrame.values`` instead (:issue:`18458`).
 - ``Series.asobject``, ``DatetimeIndex.asobject``, ``PeriodIndex.asobject`` and ``TimeDeltaIndex.asobject`` have been deprecated. Use ``.astype(object)`` instead (:issue:`18572`)
+- Grouping by a tuple of keys now emits a ``FutureWarning`` and is deprecated.
+  In the future, a tuple passed to ``'by'`` will always refer to a single key
+  that is the actual tuple, instead of treating the tuple as multiple keys (:issue:`18314`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.22.0.txt]: mention you can simply replace the tuple with a list

File: pandas/core/groupby.py [hunk 4]
```diff
@@ -2850,7 +2850,15 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
+    if tuple_as_list:
+        msg = ("Interpreting tuple 'by' as a list of keys, rather than "
+               "a single key. Use 'by={!r}' instead of 'by={!r}'. In the "
+               "future, a tuple will always mean a single key.".format(
+                   list(key), key))
```
Comment (jorisvandenbossche) [inline pandas/core/groupby.py]: the key can contain a long array or column, so not sure it is a good idea to format it like this into the message.

**Review by jorisvandenbossche — COMMENTED** [review]
The example from the docs (see top post https://github.com/pandas-dev/pandas/issues/18314#issue-274315173 for the code to reproduce) is still failing with this branch.

File: pandas/core/groupby.py (hunk 1 above)
Comment (toobaz) [reply pandas/core/groupby.py]: (Except that, as @jorisvandenbossche reminded below, ``set(key)`` could contains non-hashable objects, so this possibility should be catched)

File: pandas/core/groupby.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/groupby.py]: I thought NumPy's short repr kicked in sooner that it does. I'll fix this

File: pandas/core/groupby.py [hunk 5]
```diff
@@ -2850,7 +2851,25 @@ def _get_grouper(obj, key=None, axis=0, level=None, sort=True,
...
+    # a warning. We can assume that the user wanted a list of keys when
+    # the key is not in the index. We just have to be careful with
+    # unhashble elements of `key`. Any unhashable elements implies that
+    # they wanted a list of keys.
+    # https://github.com/pandas-dev/pandas/issues/18314
+    is_tuple = isinstance(key, tuple)
+    all_hashable = is_tuple and all(is_hashable(x) for x in key)
+
+    if is_tuple:
+        if not all_hashable or key not in obj:
```
Comment (toobaz) [inline pandas/core/groupby.py]: I'm lost. Why do you check that elements are _not_ hashable? I would have done instead
```python
if all_hashable and key not in obj and set(key).issubset(obj):
```
or (if we want to account for the to-be-deprecated possibility to index with missing keys):
```python
if all_hashable and key not in obj and set(key) & (obj):
```

Comment (toobaz) [reply pandas/core/groupby.py]: Or better - performance-wise:
```python
if key not in obj and all(is_hashable(x) for x in key) and set(key).issubset(obj):
```

Comment (TomAugspurger) [reply pandas/core/groupby.py]: This is for the case where you're grouping by non-hashable arrays like in https://github.com/pandas-dev/pandas/issues/18314#issue-274315173

In that case, don't we know that they're certainly relying on `groupby((a, b))` to be `groupby([a, b])`, so we want to warn and listify?

Comment (TomAugspurger) [reply pandas/core/groupby.py]: I do still need to handle your KeyError example.

Comment (TomAugspurger) [reply pandas/core/groupby.py]: Ahh, I see what you're doing now. Yes, that's probably better, and will make handling the KeyError easier.

**Comment by TomAugspurger** [comment]
Huh about the example:

```python
        df = pd.DataFrame(1, index=range(3), columns=pd.MultiIndex.from_product([[1, 2], [3,4]]))
        df.groupby((7, 8)).mean()
```

On master that gives me

```
Out[4]:
   1     2
   3  4  3  4
7  1  1  1  1
8  1  1  1  1

```

Is that correct? That seems like it should throw a KeyError, right?

**Comment by TomAugspurger** [comment]
OK, updated to use your suggestion @toobaz, with a slight modification so that we warn when either

1. the tuple isn't a valid key and elements are hashable and all of the elements are valid keys
2. any of the elements are not hashable

**Comment by toobaz** [comment]
Yeah, I think that's perfect (I had forgot case 2.). Two small comments:
- you could replace ``all(is_hashable(x) for x in key)`` with ``is_hashable(key)`` - but without significant performance gain, so not sure it's an improvement
- the [duplicated test](https://github.com/pandas-dev/pandas/pull/18731#discussion_r156228903) is still there

**Comment by TomAugspurger** [comment]
Thanks, fixed. Should be good to go hopefully.

**Comment by jreback** [comment]
lgtm. needs a rebase to fix conflict. merge on green.

**Pull request merged**